Simplified/OPDS/NYPLOPDSFeedFetcher.swift

```diff
@@ -0,0 +1,122 @@
+//
+//  NYPLOPDSFeedFetcher.swift
+//  Simplified
+//
+//  Created by Ernest Fan on 2022-01-09.
+//  Copyright © 2022 NYPL. All rights reserved.
+//
+
+import Foundation
+
+@objcMembers class NYPLOPDSFeedFetcher: NSObject {
+  // The the maximum attempts of books fetching allowed until we receive supported books.
+  private static let fetchSupportedBooksRetryThreshold = 15
+  
+  ///   Fetch catalog feed for given URL.
+  ///   If the returned feed does not contain supported books, it will call itself recursively
+  ///   until supported books are received or retry threshold reached.
+  ///
+  ///   - Parameter url: URL for fetching the catalog feed
+  ///   - Parameter retryCount: Current number of attempts to fetch supported books
+  ///   - Parameter completion: Always invoked at the end no matter what,
+  ///   providing an ungrouped feed object in case of success and nil otherwise.
+  class func fetchCatalogUngroupedFeed(url: URL?,
+                                       networkExecutor: NYPLNetworkExecuting,
+                                       retryCount: Int = 0,
+                                       completion: @escaping (_ feed: NYPLCatalogUngroupedFeed?) -> Void) {
+    guard let url = url else {
+      Log.error(#function, "Unable to fetch feed with empty URL.")
+      completion(nil)
+      return
+    }
+    
+    if (retryCount >= fetchSupportedBooksRetryThreshold) {
+      Log.warn(#function, "Retry threshold reached while fetching Catalog feed.")
+      completion(nil)
+      return
+    }
+    
+    fetchOPDSFeed(url: url,
+                  networkExecutor: networkExecutor,
+                  shouldResetCache: false) { feed, error in
+      guard let feed = feed,
+            feed.type == NYPLOPDSFeedType.acquisitionUngrouped else {
+        completion(nil)
+        return
+      }
+    
+      let catalogFeed = NYPLCatalogUngroupedFeed.init(opdsFeed: feed)
+      if let catalogFeed = catalogFeed,
+          catalogFeed.books.count == 0 {
+        fetchCatalogUngroupedFeed(url: catalogFeed.nextURL,
+                                  networkExecutor: networkExecutor,
+                                  retryCount: retryCount + 1,
+                                  completion: completion)
+      } else {
+        completion(catalogFeed)
+      }
+    }
+  }
+  
+  ///   Fetch OPDS feed with a GET request for the given URL.
+  ///
+  ///   - Parameter url: The URL to contact
+  ///   - Parameter shouldResetCache: Pass YES to wipe the whole cache.
+  ///   - Parameter completion: Always invoked at the end no matter what,
+  ///   providing an OPDS feed object in case of success and an dictionary containing error information otherwise.
+  class func fetchOPDSFeed(url: URL,
```

Comment:
similarly, I'd make the url param optional, since this function can be called from objc, similarly to what you did for the other refactored function.